ipa-metrics/src/label.rs

```diff
@@ -0,0 +1,170 @@
+use std::{
+    fmt::{Debug, Display, Formatter},
+    hash::{Hash, Hasher},
+};
+
+pub use Value as LabelValue;
+
+pub const MAX_LABELS: usize = 5;
+
+/// Dimension value (or label value) must be sendable to another thread
+/// and there must be a way to show it
+pub trait Value: Display + Send {
+    /// Creates a unique hash for this value.
+    /// It is easy to create collisions, so better avoid them,
+    /// by assigning a unique integer to each value
+    fn hash(&self) -> u64;
+
+    /// Creates an owned copy of this value. Dynamic dispatch
+    /// is required, because values are stored in a generic store
+    /// that can't be specialized for value types.
+    fn boxed(&self) -> Box<dyn LabelValue>;
+}
+
+impl LabelValue for u32 {
+    fn hash(&self) -> u64 {
+        u64::from(*self)
+    }
+
+    fn boxed(&self) -> Box<dyn LabelValue> {
+        Box::new(*self)
+    }
+}
+
+#[derive()]
+pub struct Label<'lv> {
+    pub name: &'static str,
+    pub val: &'lv dyn Value,
+}
+
+impl Label<'_> {
+    #[must_use]
+    pub fn to_owned(&self) -> OwnedLabel {
+        OwnedLabel {
+            name: self.name,
+            val: self.val.boxed(),
+        }
+    }
+}
+
+impl Debug for Label<'_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Label")
+            .field("name", &self.name)
+            .field("val", &format!("{}", self.val))
+            .finish()
+    }
+}
+
+impl Hash for Label<'_> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        state.write(self.name.as_bytes());
+        state.write_u64(self.val.hash());
+    }
+}
+
+impl PartialEq for Label<'_> {
+    fn eq(&self, other: &Self) -> bool {
+        // name check should be fast - just pointer comparison.
+        // val check is more involved with dynamic dispatch, so we can consider
+        // making label immutable and storing a hash of the value in place
+        self.name == other.name && self.val.hash() == other.val.hash()
```

Comment:
Implementing `eq` by comparing hashes seems dangerous. Maybe add an `eq` function to the `(Label)Value` trait so that value types can implement `eq` in terms of the hash only when it is safe to do so?

Reply:
Object safety is on our way to provide it :( I added a note to `hash` function that it will be used to check uniqueness. Happy to do more if there is a reasonable way to deal with it.